Flux Correct Conservation Bugfix

mabruzzo opened the pull request:
This includes a bugfix to `MethodFluxCorrect` to make sure that quantities like `total_energy`, `velocity_x`, `velocity_y`, and `velocity_z` are converted to a conserved quantity while performing flux corrections. This addresses the problems outlined in Issue #56 pertaining to `MethodFluxCorrect`.

bwoshea commented:
@gregbryan  and @forrestglines  -- I assigned you to review this (short, easy to review) PR.  Would you mind doing that when you get a moment?  Thanks!

mabruzzo commented:
I also added some tests to try to make sure we don't break this going forward. Currently the test runs an inclined entropy wave in a simulation with SMR (since SMR currently conserves data better than AMR)

I just wanted to make 2 notes about the tests that I have added:
1. These tests only checks how well density is conserved. It would probably be good to extend the test to also check the conservation of the other quantities.
2. The level of conservation for the single-precision test that I added is lower than I was expected. While density is conserved to 7 digits, everything else is only conserved about 5 digits. In comparison, for the double precision test, every quantity is conserved to a similar precision (I think about ~16 digits).

I might try to address the first point, but I don't have plans to address the second one (maybe I'll make a new issue just so that we have a record of it).

mabruzzo commented:
Wrapping this up this took more effort than I was expecting, but this is ready for review

The remaining changes largely addressed the following concern (that I previously raised):
>     1. These tests only checks how well density is conserved. It would probably be good to extend the test to also check the conservation of the other quantities.
To address this, I modified `MethodFluxCorrect` and the input files to now check the conservation of all of the fields (addressing the above concern that I had highlighted). 

I've also updated the website documentation to reflect all changes introduced in this PR.

---
I also looked slightly more deeply into the other concern that I had mentioned:
>     2. The level of conservation for the single-precision test that I added is lower than I was expected. While density is conserved to 7 digits, everything else is only conserved about 5 digits. In comparison, for the double precision test, every quantity is conserved to a similar precision (I think about ~16 digits).

I can get a little more specific about the numbers that I'm seeing (Some of these numbers are taken from the input parameter files where I list the expected values so the numbers given here might be ~0.1 smaller than what the simulation reports).


|                         | `CELLO_PREC=single` | ``CELLO_PREC=double`` |
|---------------------|-----------------------------------|------------------------------------|
| "density"          |                                 ~7.1 |                                 ~16.4 |
| "total_energy" |                                 ~4.9 |                                  ~16.5 |
| "velocity_x"     |                                 ~4.9 |                                     ~16 |
| "velocity_y"     |                                 ~4.9 |                                  ~15.9 |
| "velocity_z"     |                                 ~4.9 |                                  ~15.9 |

Due to a failure of the tests on CircleCi, I checked to see how much my results change when I compiled the Enzo-E on a different machine (with a different version of GCC). The number of conserved digits on that machine are reported below:

|                         | ``CELLO_PREC=single`` | ``CELLO_PREC=double`` |
|---------------------|-----------------------------------|------------------------------------|
| "density"          |                                 ~6.3 |                                     ~16 |
| "total_energy" |                                  ~5.4 |                                  ~15.8 |
| "velocity_x"     |                                  ~5.3 |                                  ~15.9 |
| "velocity_y"     |                                  ~5.3 |                                  ~15.9 |
| "velocity_z"     |                                  ~5.3 |                                  ~15.7 |

The results from the second machine make me conclude that there probably isn't an issue with flux conservation when Enzo-E is compiled in single precision (Frankly, I'm a little surprised about how much we see the numbers change between machines)

bwoshea commented:
@mabruzzo thanks for the update on this - I will work on the reviewers!

forrestglines reviewed (approved):
Looks good! I only have a couple minor suggestions.

forrestglines commented:
> To address this, I modified `MethodFluxCorrect` and the input files to now check the conservation of all of the fields (addressing the above concern that I had highlighted).

This is good - testing all the fields is definitely a requirement for these tests to be complete.

> The results from the second machine make me conclude that there probably isn't an issue with flux conservation when Enzo-E is compiled in single precision (Frankly, I'm a little surprised about how much we see the numbers change between machines)

The conservation difference in double vs. single is extremely strange, especially for different machines both with GCC. I would expect those differences to arise for different compilers (i.e. GCC vs. Intel) or different CPU vendors. Since the doubles are working well, I don't think this issue is worth spending much time to examine right now.

mabruzzo commented on input/FluxCorrect/inclined_contact_ppm.incl:
```diff
@@ -0,0 +1,94 @@
+	      "total_energy", "internal_energy"];
+
+      gamma = 1.6666666666666667;
+      ghost_depth = 4; # needs to be 4 for flux corrections
```
PPM uses third order interpolation (I always mix this up myself). According to the Enzo Paper, it requires 3 ghost zones. We need to round up to 4 zones since AMR requires an even number of ghost zones.

mabruzzo commented on src/Cello/problem_MethodFluxCorrect.cpp:
```diff
@@ -203,120 +227,290 @@ void MethodFluxCorrect::compute_continue_sum_fields
+  } // rank >= 3
+}
+
+static void divide_by_density_(const cello_float * const input_array,
```
Good point!

I choose a slightly different name to be more explicit, ``divide_flux_corrected_cells_by_density_``.

mabruzzo commented on doc/source/param/method.incl:
```diff
@@ -74,6 +75,54 @@ flux_correct
+           min_digits = ["density", 7.1,
+                         "velocity_x", 4.9];
+        }
+     }
```
Yes (assuming that velocity_x is in the ``"make_field_conservative"`` field group). I edited this to try to clear up the confusion.

mabruzzo commented on src/Cello/problem_MethodFluxCorrect.cpp:
```diff
@@ -203,120 +227,290 @@ void MethodFluxCorrect::compute_continue_sum_fields
+          for (int iy=0; iy<ny; iy++) {
+            for (int ix=0; ix<nx; ix++) {
+              int i=ix + mx*(iy + my*iz);
+              temp_cons_array[i] = old_rho_array[i]*field_array[i];
```
Exactly! In other words, to apply flux corrections for cells adjacent to the x-face:

- ρ_new = ρ_old + (dt/dx)*δF_x
- v_x,new = (ρ_old*v_x,old + (dt/dx)*δF_x)/ρ_new

Note: In Enzo-E, the factor of (dt/dx) is absorbed into the coarse and fine fluxes (I think this might help when you have adaptive time stepping in a cosmological simulation)

gregbryan commented on doc/source/param/method.incl:
```diff
@@ -74,6 +75,64 @@ flux_correct
+----
+
+:Parameter:  :p:`Method` : :p:`flux_correct` : :p:`min_digits`
+:Summary: :s:`Number of digits to be conserved by fields in tests`
```
```suggestion
:Summary: :s:`Number of digits expected to be conserved by fields in tests`
```
Suggest adding "expected" just to make this clearer.

gregbryan commented on src/Cello/problem_MethodFluxCorrect.cpp:
```diff
@@ -203,120 +227,291 @@ void MethodFluxCorrect::compute_continue_sum_fields
+
+    // Allocate scratch space (if not already allocated necessary). It's
+    // technically possible to do all of these operations in-place, (without
+    // scratch space), but that get's complex (you need to be very careful
```
```suggestion
    // scratch space), but that gets complex (you need to be very careful
```

gregbryan reviewed (approved):
This looks good to me!

mabruzzo commented:
@pgrete Do you intend to review this? There's absolutely no rush for you to do so. I just wanted to confirm your intentions (otherwise, this could probably be merged in).

pgrete commented:
I'm actually not sure how I ended up as a reviewer here as I think two reviews are fine.
So I'm pulling the trigger now.
